src/backend/distributed/connection/remote_commands.c

```diff
@@ -75,6 +75,90 @@ ForgetResults(MultiConnection *connection)
 
 
 /*
+ * NonblockingForgetResults clears a connection from pending activity if doing
+ * so does not require network IO. Returns true if successful, false
+ * otherwise.
+ */
+bool
+NonblockingForgetResults(MultiConnection *connection)
+{
+	PGconn *pgConn = connection->pgConn;
+	bool wasNonblocking = PQisnonblocking(pgConn);
+
+	if (PQstatus(pgConn) != CONNECTION_OK)
+	{
+		return false;
+	}
+
+	while (true)
+	{
+		PGresult *result = NULL;
+
+		CHECK_FOR_INTERRUPTS();
+
+		if (!wasNonblocking)
```

Comment:
>  Ideally, we would just set our sockets to non-blocking mode, though I'm guessing that will mess up copy. 

Yes, it'd currently upset copy, and citus-tools.  Once we've converted them to be interruptible, we should be good to go however.

> To simplify the code, could we maybe add a flag in MultiConnection that says whether the connection is blocking so we can get rid of the wasNonBlocking checks?

Wouldn't that mean we'd have to reset it in a bunch of places that are currently unrelated?  I think I'd rather get rid of the last few places that require blocking sockets and then always put them into nonblocking mode.